File: src/primer/resolver.py
import re
import httpx
from primer.models import Source


def resolve_url(source: Source) -> str:
    """Resolve a source's URL pattern to a concrete URL by finding the latest date."""
    if source.url and not source.url_pattern:
        return source.url
    if not source.url_pattern:
        raise ValueError(f"Source {source.id} has no url or url_pattern")

    pattern = source.url_pattern
    base_url = pattern.rsplit("/", 1)[0] + "/"
    filename_pattern = pattern.rsplit("/", 1)[1]

    # Build regex: replace {date} with capturing group for YYYY-MM
    regex_str = re.escape(filename_pattern).replace(r"\{date\}", r"(\d{4}-\d{2})")
    regex = re.compile(regex_str)

    response = httpx.get(base_url, follow_redirects=True, timeout=30)
    response.raise_for_status()

    matches = []
    for match in regex.finditer(response.text):
        date = match.group(1)
        filename = match.group(0)
        matches.append((date, filename))

    if not matches:
        raise ValueError(f"No matching files found for {source.id} at {base_url}")

    matches.sort(key=lambda x: x[0])
    latest_date, latest_filename = matches[-1]
    return base_url + latest_filename


def resolve_all(sources: list[Source]) -> dict[str, str]:
    """Resolve all sources to concrete URLs. Returns {source_id: url}."""
    resolved = {}
    for s in sources:
        try:
            resolved[s.id] = resolve_url(s)
        except Exception as e:
            resolved[s.id] = f"ERROR: {e}"
    return resolved

File: host-cli/internal/resolver/resolver.go
// Package resolver resolves download URLs from static values or date-pattern templates.
package resolver

import (
	"fmt"
	"io"
	"net/http"
	"regexp"
	"sort"
	"strings"
	"time"
)

// Resolve returns a concrete download URL. If staticURL is non-empty it is
// returned directly. Otherwise urlPattern is treated as a template containing
// a {date} placeholder: the base directory is fetched via HTTP and the
// filename with the lexicographically latest date match is selected.
func Resolve(staticURL, urlPattern string) (string, error) {
	if staticURL != "" {
		return staticURL, nil
	}
	if urlPattern == "" {
		return "", fmt.Errorf("no URL or pattern")
	}

	// Split at last "/" into base URL and filename pattern.
	lastSlash := strings.LastIndex(urlPattern, "/")
	if lastSlash < 0 {
		return "", fmt.Errorf("pattern %q has no path separator", urlPattern)
	}
	baseURL := urlPattern[:lastSlash]
	filenamePattern := urlPattern[lastSlash+1:]

	// Build regex from the filename pattern, replacing {date} with a capture group.
	datePattern := strings.ReplaceAll(filenamePattern, "{date}", `(\d{4}-\d{2})`)
	fileRe, err := regexp.Compile("^" + datePattern + "$")
	if err != nil {
		return "", fmt.Errorf("compiling pattern: %w", err)
	}

	// Fetch the directory listing.
	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Get(baseURL)
	if err != nil {
		return "", fmt.Errorf("fetching %s: %w", baseURL, err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(io.LimitReader(resp.Body, 10<<20)) // 10 MB limit
	if err != nil {
		return "", fmt.Errorf("reading response: %w", err)
	}

	// Extract all href values.
	hrefRe := regexp.MustCompile(`href="([^"]+)"`)
	matches := hrefRe.FindAllStringSubmatch(string(body), -1)

	// Match hrefs against the filename pattern and collect (date, filename) pairs.
	type entry struct {
		date     string
		filename string
	}
	var entries []entry
	for _, m := range matches {
		href := m[1]
		sub := fileRe.FindStringSubmatch(href)
		if sub == nil {
			continue
		}
		entries = append(entries, entry{date: sub[1], filename: href})
	}

	if len(entries) == 0 {
		return "", fmt.Errorf("no files matching pattern %q in %s", filenamePattern, baseURL)
	}

	// Sort by date string lexicographically and pick the latest.
	sort.Slice(entries, func(i, j int) bool {
		return entries[i].date < entries[j].date
	})

	latest := entries[len(entries)-1]
	return baseURL + "/" + latest.filename, nil
}
